### PR TITLE
Refactor prompt processor

### PR DESCRIPTION
**Pull request by bennyguo**
Changes:

- Prompt processors now return an instance of `PromptProcessorOutput`, which contains text embeddings, view-dependent text embeddings, direction configs, and a function to get text embeddings from camera settings. With this change, we can more easily implement VSD and Perp-Neg.
- `view_dependent_prompting` is now an option for guidance, not prompt processors.
- The returned dict of guidance now contains several items with keys starting with `loss_`. These items will be treated as loss terms, multiplied with the corresponding `lambda_`, and added to the final loss. With this change, we can easily adapt VSD to other systems, simply changing the guidance and loss lambdas in the configuration file.
- The LoRA of ProlificDreamer takes view-independent text embeddings instead of view-dependent ones.

### CODE REVIEW

File: threestudio/data/uncond.py [hunk 1]
```diff
@@ -217,10 +217,14 @@ def collate(self, batch) -> Dict[str, Any]:
...
         up = F.normalize(torch.cross(right, lookat), dim=-1)
-        c2w: Float[Tensor, "B 3 4"] = torch.cat(
+        c2w3x4: Float[Tensor, "B 3 4"] = torch.cat(
             [torch.stack([right, up, -lookat], dim=-1), camera_positions[:, :, None]],
             dim=-1,
         )
+        c2w: Float[Tensor, "B 4 4"] = torch.cat(
+            [c2w3x4, torch.zeros_like(c2w3x4[:, :1])], dim=1
+        )
+        c2w[:, 3, 3] = 0.0
```
Comment (thuliu-yt16) [inline threestudio/data/uncond.py:227]: remove this line

File: threestudio/data/uncond.py [hunk 2]
```diff
@@ -309,10 +313,15 @@ def __init__(self, cfg: Any, split: str) -> None:
...
+        print(c2w.shape, c2w3x4.shape)
```
Comment (thuliu-yt16) [inline threestudio/data/uncond.py:324]: remove the two lines

File: threestudio/models/guidance/stable_diffusion_vsd_guidance.py [hunk 3]
```diff
@@ -398,32 +410,63 @@ def get_latents(
             latents = self.encode_images(rgb_BCHW_512)
         return latents
 
-    def forward(
+    def __call__(
```
Comment (thuliu-yt16) [inline threestudio/models/guidance/stable_diffusion_vsd_guidance.py:413]: just leaving it as forward is ok

File: threestudio/models/guidance/stable_diffusion_vsd_guidance.py [hunk 4]
```diff
@@ -398,32 +410,63 @@ def get_latents(
...
         # reparameterization trick
         # d(loss)/d(latents) = latents - target = latents - (latents - grad) = grad
         target = (latents - grad).detach()
-        vsd_loss = 0.5 * F.mse_loss(latents, target, reduction="sum") / batch_size
+        loss_vsd = 0.5 * F.mse_loss(latents, target, reduction="sum") / batch_size
 
-        lora_loss = self.train_lora(latents, text_embeddings, mvp_mtx)
+        loss_lora = self.train_lora(
+            latents, text_embeddings, text_embeddings_inp, camera_condition
+        )
```
Comment (thuliu-yt16) [inline threestudio/models/guidance/stable_diffusion_vsd_guidance.py:466]: remove the useless text_embeddings in the parameters

File: threestudio/models/guidance/stable_diffusion_vsd_guidance.py [hunk 5]
```diff
@@ -398,32 +410,63 @@ def get_latents(
...
+            elevation,
+            azimuth,
+            camera_distances,
+            view_dependent_prompting=self.cfg.view_dependent_prompting,
+        )
+
+        # input text embeddings, view-independent
+        text_embeddings_inp = prompt_utils.get_text_embeddings(
+            elevation, azimuth, camera_distances, view_dependent_prompting=False
+        )
```
Comment (thuliu-yt16) [inline threestudio/models/guidance/stable_diffusion_vsd_guidance.py:441]: would be better to rename text_embeddings -> text_embeddings_vd, text_embeddings_inp -> text_embeddings

**Review by thuliu-yt16 — CHANGES_REQUESTED** [review]
see the comments

File: threestudio/data/uncond.py (hunk 1 above)
Comment (bennyguo) [reply threestudio/data/uncond.py:227]: right

File: threestudio/models/guidance/stable_diffusion_vsd_guidance.py (hunk 3 above)
Comment (bennyguo) [reply threestudio/models/guidance/stable_diffusion_vsd_guidance.py:413]: right, I forget about this when copy and paste

File: threestudio/models/guidance/stable_diffusion_vsd_guidance.py (hunk 5 above)
Comment (bennyguo) [reply threestudio/models/guidance/stable_diffusion_vsd_guidance.py:440]: done

File: threestudio/models/guidance/stable_diffusion_vsd_guidance.py (hunk 4 above)
Comment (bennyguo) [reply threestudio/models/guidance/stable_diffusion_vsd_guidance.py:463]: done
